src/v/raft/tests/state_machine_manager_tests.cc

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "raft/tests/raft_fixture.h"
+#include "test_utils/randoms.h"
+
+using namespace raft;
+
+ss::logger test_logger{"state_machine_test_logger"};
+
+// Stm that causes a lag of _next in stm_manager.
+class slow_stm : public state_machine_base {
+public:
+    static constexpr std::string_view name = "slow_stm";
+
+    ss::future<> apply(const model::record_batch&) override {
+        return ss::sleep(5ms);
+    }
+
+    ss::future<> apply_raft_snapshot(const iobuf&) override {
+        return ss::now();
+    };
+
+    ss::future<iobuf>
+    take_snapshot(model::offset last_included_offset) override {
+        co_return iobuf{};
+    };
+
+    size_t get_local_state_size() const override { return 0; };
+
+    ss::future<> remove_local_state() override { return ss::now(); };
+
+protected:
+    ss::future<> start() override { return ss::now(); }
+    ss::future<> stop() override { return ss::now(); }
+};
+
+// stm that results in a background apply. It works by failing
+// the first apply, resulting in a background application fiber
+// and slowing down further apply calls causing the background
+// fiber to not exit.
+class failing_stm : public state_machine_base {
+public:
+    static constexpr std::string_view name = "failing_stm";
+
+    failing_stm() {}
+
+    ss::future<> apply(const model::record_batch&) override {
+        if (_first) {
+            _first = false;
+            return ss::make_exception_future(std::runtime_error("failure"));
+        }
+        return ss::sleep(20ms);
+    }
+
+    ss::future<> apply_raft_snapshot(const iobuf&) override { co_return; };
+
+    ss::future<iobuf>
+    take_snapshot(model::offset last_included_offset) override {
+        co_return iobuf{};
+    };
+
+    size_t get_local_state_size() const override { return 0; };
+
+    ss::future<> remove_local_state() override { return ss::now(); };
+
+protected:
+    ss::future<> start() override { return ss::now(); }
+    ss::future<> stop() override { return ss::now(); }
+
+private:
+    bool _first = true;
+};
+
+class failing_stm_fixture
+  : public stm_raft_fixture<failing_stm_fixture, failing_stm, slow_stm> {
+public:
+    failing_stm_fixture() = default;
+
+    stm_shptrs_t create_stms(
+      state_machine_manager_builder& builder, raft_node_instance& node) {
+        return {
+          builder.create_stm<failing_stm>(), builder.create_stm<slow_stm>()};
+    }
+};
+
+TEST_F_CORO(failing_stm_fixture, snapshot_and_background_apply_test) {
```

Comment:
can we add it to the `stm_manager_test.cc` ?

Reply:
I ported the test to `stm_manager_test.cc`, it can reproduce the assert in the issue.